tools/cbmc/proofs/HTTP/global_state_HTTP.c

```diff
@@ -0,0 +1,201 @@
+#define MAX_DATA_SIZE 1000
+
+/* Implementation of safe malloc which returns NULL if the reuested size is 0.*/
+void *safeMalloc(size_t xWantedSize) {
+  if(xWantedSize == 0) {
+    return NULL;
+  }
+  uint8_t byte;
+  return byte ? malloc(xWantedSize) : NULL;
+}
+
+/* This modeling is required because otherwise CBMC takes
+too much time if assigned to minimumBufferSizes. */
+typedef struct _responseHandle
+{
+    struct _httpsResponse RespHandle;
+    char data[MAX_DATA_SIZE];
+} _resHandle_t;
+
+typedef struct _requestHandle
+{
+    struct _httpsRequest ReqHandle;
+    char data[MAX_DATA_SIZE];
+} _reqHandle_t;
+
+typedef struct _connectionHandle
+{
+    struct _httpsConnection pConnHandle;
+    char data[MAX_DATA_SIZE];
+} _connHandle_t;
+
+/* Models the third party HTTP Parser. */
+size_t http_parser_execute (http_parser *parser,
+                            const http_parser_settings *settings,
+                            const char *data,
+                            size_t len) {
+  size_t ret;
+  _httpsResponse_t *_httpsResponse = (_httpsResponse_t *)(parser->data);
+  /* nondet_bool is required to get coverage. */
+  _httpsResponse->foundHeaderField = nondet_bool();
+  _httpsResponse->parserState = PARSER_STATE_BODY_COMPLETE;
+  return ret;
+}
+
+IotNetworkError_t IotNetworkInterfaceCreate( void * pConnectionInfo,
+               void * pCredentialInfo,
+               void * pConnection ) {
+    *(char **)pConnection = malloc(1); /* network connection is opaque */
+    IotNetworkError_t error;
+    return error;
+}
+
+size_t IotNetworkInterfaceSend( void * pConnection,
+           const uint8_t * pMessage,
+           size_t messageLength ) {
+    size_t size;
+    __CPROVER_assume(size <= messageLength);
+    return size;
+}
+
+IotNetworkError_t IotNetworkInterfaceClose( void * pConnection ) {
+    IotNetworkError_t error;
+    return error;
+}
+
+size_t IotNetworkInterfaceReceive( void * pConnection,
+        uint8_t * pBuffer,
+        size_t bytesRequested ) {
+    uint8_t byte;
+    __CPROVER_array_copy(pBuffer,&byte);
```

Comment:
This is a standard trick, but I'd document for the newcomer that this intrinsic copies a source to a destination, and fills the remainder of the destination with unconstrained data.